client/src/templates/Challenges/codeally/show.tsx

```diff
@@ -176,6 +165,45 @@ class ShowCodeAlly extends Component<ShowCodeAllyProps> {
     }
   };
 
+  openGitpod = (userToken?: string) => {
+    const {
+      data: {
+        challengeNode: {
+          challenge: { url, coderoadTutorial }
+        }
+      }
+    } = this.props;
+
+    const gitpodDomain = `https://gitpod.io/?autostart=true#CODEROAD_TUTORIAL_URL=${coderoadTutorial},CODEROAD_DISABLE_RUN_ON_SAVE=true`;
+    const tokenEnv = userToken ? `,CODEROAD_WEBHOOK_TOKEN=${userToken}` : '';
+    const gitpodUrl = `${gitpodDomain}${tokenEnv}/${url}`;
+
+    window.open(gitpodUrl, '_blank');
+  };
```

Comment:
I really do not like this current implementation, but am fine with it for the initial release. I do not think this should be a `window.open` API. I think this should be more like an `a` element whereby it also shows the link Campers will be taken to when they hover over the button.

At least in all Chromium browsers, hovering over a button shows nothing, but hovering over `<a href="https://freecodecamp.org"></a>` shows `https://freecodecamp.org` in the bottom left-hand corner.

Reply:
Agreed. All the logic in `startCourse` should be inside `useEffect`.

Reply:
Just to clarify: I'm also fine with this for now (moving to functional components would be worthy of a PR on its own, let alone changing the feature).